### ISSUE TITLE
VIS imaging tutorials are too confusing

### ISSUE THREAD

**Issue by AdamDimech** [danforthcenter/plantcv#318]
I am a beginner to PlantCV and am struggling with the published tutorials for VIS imaging. I know that this issue has been raised before, but it does not satisfactorily help a beginner to python and PlantCV. I am looking to analyse images similar to those in the tutorial.

My starting point was to follow the [VIS tutorial](https://plantcv.readthedocs.io/en/latest/vis_tutorial/)., which results in the [following code](https://gist.github.com/AdamDimech/a863cc3b04d76e922db630276366e76a). (All of the pieces were copied and pasted into one file in sequence). **Issue 1:** Why is there no summary file at the end that one can copy and paste from?

Trial-and-error has shown me that this isn't actually Python-ready and there is no mention of this in the relevant tutorial page. I have figured out that this needs to be saved as a .ipynb and then converted to .py via `jupyter nbconvert --to python VIS_tutorial.ipynb` ([Source](https://plantcv.readthedocs.io/en/latest/jupyter/)). **Issue 2**: The tutorial omits the requirement to add `%matplotlib inline` and some other headers as well. Anyway, following this, a [new .ipynb is created](https://gist.github.com/AdamDimech/830d1de93d8a264322310539d2253fc0). This is then converted to a .py ([see Gist for code](https://gist.github.com/AdamDimech/1f1a9fc20c2c15e18f0401c52f8b871c)).

**Issue 3**: Running the command `VIS_tutorial2.py -i original_image.jpg -o ./output-images -r results.txt 
-w -D 'print'` - [as shown in the tutorial](https://plantcv.readthedocs.io/en/latest/vis_tutorial/) - triggers an error:
```
"usage: VIS_tutorial2.py [-h] -i IMAGE [-o OUTDIR] [-r RESULT] [-w WRITEIMG]
                        [-D DEBUG]
VIS_tutorial2.py: error: argument -w/--writeimg: expected one argument" 
```
Okay, fix and run again: `VIS_tutorial2.py -i original_image.jpg -o ./output-images -r results.txt -w test.jpg -D 'print'`. 
This time the error reads: 

```
File "G:\Path\To\PlantCV\VIS_tutorial2.py", line 140, in <module>
    main()
  File "G:\Path\To\PlantCV\VIS_tutorial2.py", line 51, in main
    s_thresh = pcv.threshold.binary(s, 85, 255, 'light')
NameError: name 's' is not defined
```
I am aware that some working demo code has been published ([Code](https://github.com/danforthcenter/plantcv-terra/blob/master/notebooks/demo_notebook.ipynb), [Reference](https://github.com/danforthcenter/plantcv/issues/198#issuecomment-341483345)). Going through the steps again to convert the .ipynb to .py in Jupyter ([code on GitHub](https://gist.github.com/AdamDimech/920435a23dfa203f513fd6f038026674)) results in a different error: 

```
  File "G:\Path\To\PlantCV\Demo.py", line 109, in <module>
    img, path, filename = pcv.readimage(filename=image, debug=debug)
AttributeError: module 'plantcv' has no attribute 'readimage'
```
I don't even know where to go with that error.

I am a bit flabbergasted that none of this works, honestly. I am not a highly-experienced programmer but not a novice to command-line either and this is just baffling.

**In summary**: My issue is that the tutorials seem to be non-sequential in parts, omit key instructions and as far as I can tell, there is no obvious place where one can simply get PlantCV up-and-running and analyse a single image as a "proof-of-concept" to demonstrate simple functionality. I have provided my steps in detail to hopefully show you how the documentation presents to newbies.

I am running PlantCV on a Windows 10 machine via Anaconda Navigator.

**Comment by nfahlgren** [comment]
For issue 1 I created issue #319, we can add links to complete scripts for the tutorials.

For issue 2, the main VIS, NIR, VIS/NIR, and Multi-plant tutorials are written as scripts, not Jupyter notebooks, so `%matplotlib inline` is not required, neither is adding them to a notebook first and then using `nbconvert` to convert them back to scripts. You can do that of course, but it's definitely not required. As @HaleySchuhl mentioned, she has taken the tutorials and put them into Jupyter notebooks in the Binder environment so that people can use them directly with example data.

When people develop their own image analysis workflows we recommend building them in Jupyter first, then converting them to scripts because it's easier to build and visualize in Jupyter. That's just our preference though, it's also perfectly valid to develop in script form.

For issue 3, in `options`, the line `parser.add_argument("-w","--writeimg", help="write out images.", default=False)` should be `parser.add_argument("-w","--writeimg", help="write out images.", default=False, action="store_true")`. We will need to fix that when we update for issue #319.

The missing `s` variable is because you missed this line `s = pcv.rgb2gray_hsv(img, 's')` between the read image and threshold the saturation image steps. When we add the full scripts to the tutorials for issue #319 it should help make it easier to use the code without going through each block in the documentation.

The error `AttributeError: module 'plantcv' has no attribute 'readimage'` is because the examples you found are out of date. They all use `import plantcv as pcv`. This works but does not import the PlantCV main library, instead imports should look like `from plantcv import plantcv as pcv`. This will allow us to build on add-on packages as discussed in issue #219. A full set of documentation to update older scripts/notebooks to the current program interface can be found at https://plantcv.readthedocs.io/en/latest/updating/#updating-to-v3

**Comment by AdamDimech** [comment]
Thank you both for your very helpful responses and code corrections. Updating the tutorials will be a big help, but what finally got me up-and-running was the [Binder Interactive Tutorial](https://mybinder.org/v2/gh/danforthcenter/plantcv-binder.git/master?filepath=index.ipynb). I have now created a functional script that will work with one image. The next step is to build a pipeline off that. If I encounter any more problems that relate to the instructions, I will update this thread.

**Comment by nfahlgren** [comment]
Thanks @AdamDimech, let us know if you have other suggestions for improvements.

**Comment by AdamDimech** [comment]
When I attempted to parallelise my optimised pipeline, I received the following error after following [the tutorial](https://plantcv.readthedocs.io/en/latest/pipeline_parallel/#running-plantcv-pipelines-over-a-flat-directory-of-images) for running analysis over a flat directory of images: 

`Traceback (most recent call last): File "C:/path/to/VIS_pipeline.1.2.py", line 354, in <module> main() File "C:/path/to/VIS_pipeline.1.2.py", line 246, in main obj, mask = pcv.object_composition(img, roi_objects, hierarchy3) File "C:\path\to\Anaconda\envs\plantcv\lib\site-packages\plantcv\plantcv\object_composition.py", line 37, in object_composition r, g, b = cv2.split(ori_img) ValueError: too many values to unpack (expected 3)`

This issue was actually addressed with [Issue #255](https://github.com/danforthcenter/plantcv/issues/255) and the solution there worked for me when I read-in the image as RGB:

`img, path, filename = pcv.readimage(args.image, mode="rgb")`

This code change needs to be made to the [VIS Tutorial](https://plantcv.readthedocs.io/en/latest/vis_tutorial/).

Additionally, on the VIS Tutorial page there is a [link to the complete script](https://plantcv.readthedocs.io/en/latest/vis_tutorial/scripts/vis_script.md) which returns a 404.

**Comment by HaleySchuhl** [comment]
@nfahlgren I added `mode='rgb'` to the VIS tutorial and noted that this is needed for RGBA images within my current branch. The links work within PyCharm, so I'll see if adding them to the table of contents will resolve the 404 since it seems to try redirecting to the table of contents when you click the link online.

**Comment by AdamDimech** [comment]
@nfahlgren I agree that it would be preferable that [plantcv.readimage](https://plantcv.readthedocs.io/en/latest/read_image/) defaults to RGB because, as you say, there is seldom a case where the alpha channel contains any information. Perhaps `mode="rgba"` should be added as an option that users can use if it is ever required.

@HaleySchuhl Thank you.

**Comment by HaleySchuhl** [comment]
@AdamDimech I think rather than changing the default from 'native' to 'rgb' in `pcv.readimage` it would be better to look at the shape of the image. We could default to 'rgb' when 4 channels are found and add the `mode='rgba'` as you suggested. This way people can still read in grayscale images under default settings.
